src/v/datalake/translation/tests/scheduler_tests.cc

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/logger.h"
+#include "datalake/translation/tests/scheduler_fixture.h"
+#include "test_utils/randoms.h"
+
+#include <seastar/core/sleep.hh>
+
+using namespace datalake::translation::scheduling;
+
+class many_concurrent_translators : public scheduler_fixture {
+public:
+    std::unique_ptr<scheduling_policy> make_scheduling_policy() override {
+        return scheduling_policy::make_default(30, task_time_quota);
+    }
+};
+
+TEST_F_CORO(many_concurrent_translators, test_concurrent_add_remove) {
+    std::vector<translator_id> added;
+    auto add_one = [this, &added]() {
+        auto translator = make_random_translator();
+        auto ntp = translator->id();
+        return _scheduler->add_translator(std::move(translator))
+          .then([&added, ntp = std::move(ntp)](bool success) {
+              if (!success) {
+                  return ss::make_exception_future(
+                    std::runtime_error("translator addition failed"));
+              }
+              added.push_back(ntp);
+              return ss::now();
+          })
+          .then([]() { return ss::sleep(100ms); });
+    };
+
+    auto remove_one = [this, &added]() {
+        if (added.empty()) {
+            return ss::now();
+        }
+        auto idx = random_generators::get_int<size_t>(added.size() - 1);
+        auto to_remove = added[idx];
+        added.erase(added.begin() + idx);
+        return ss::do_with(
+          std::move(to_remove), [this](translator_id& to_remove) {
+              return _scheduler->remove_translator(to_remove).then(
+                []() { return ss::sleep(150ms); });
+          });
+    };
+
+    auto end_time = clock::now() + 120s;
+    auto add_f = ss::do_until(
+      [end_time] { return clock::now() >= end_time; },
+      [&]() { return add_one(); });
+    auto remove_f = ss::do_until(
+      [end_time] { return clock::now() >= end_time; },
+      [&]() { return remove_one(); });
+
+    co_await ss::when_all_succeed(std::move(add_f), std::move(remove_f));
```

Comment:
This is a great fuzz test, but is there anything we can actually assert? For example making sure we adhere to the policy and no resource is oversubscribed?

Reply:
I added some basic validation.. I was thinking we could add more interesting validations along with the actual scheduling policy patch.